File: bin/build_time_lapse.pl
#!/usr/bin/perl

use strict;
use warnings;
use Fcntl qw(:flock);
use Time::ParseDate;

use DateTime;
use SC6::Cam::General;
use SC6::Cam::Config;
use SC6::Cam::Sun;
use Getopt::Long;
use Data::Dumper;
use IO::Select;
use IPC::Open3;
use Symbol 'gensym';

$|++;

my $force = 0;
my $mode = "prod";
my $dryrun = 0;
my $sleep_time = 30;

my $c = new SC6::Cam::Config("/usr/local/cam/conf/config.yml");
our $config = $c->getConfig();
our $debug = $c->getDebug();
my $date;
my $no_push;
my $silent = 0;
my $trickle = 0;
my $trickle_cmd = "trickle -s -u 200";
my $keep = 0;
my $out = "";
my $directory;   # overrides the date
my $fps;   # overrides fps

my $result = GetOptions (  "n|dry-run" => \$dryrun,
                        "f|force"  => \$force,
                        "h|help"  => \&usage,
                        "k|keep"  => \$keep,
                        "m|mode=s"  => \$mode,
                        "t|date=s"  => \$date,
                        "D|directory=s"  => \$directory,
                        "np|no-push"  => \$no_push,
                        "trickle"  => \$trickle,
                        "fps=f"  => \$fps,
                        "silent"  => \$silent,
                        "d|debug+"  => \$debug);

if ( ! $result ) {
    usage();
    exit;
}

# debug trumps silent
if ( $debug && $silent ) {
    $silent = 0;
}

unless (flock(DATA, LOCK_EX|LOCK_NB)) {
    if ( ! $silent ) {
        print "$0 is already running. Exiting.\n";
    }
    exit(1);
}

my $dt;
my $date_from_args = 0;
if ( $date ) {
    print $date, "\n" if ( $debug );
    my ($seconds, $error) = parsedate($date);
    if ( not $seconds ) {
	print "Error can't parse date: $date : $error\n";
        exit(1);
    }
    else {
	print $seconds, "\n" if ( $debug );
        $dt = DateTime->from_epoch(  epoch => $seconds, time_zone => $config->{'General'}->{'Timezone'} );
    }
    $force = 1;
    $date_from_args = 1;
}
else {
    $dt = DateTime->now(  time_zone => $config->{'General'}->{'Timezone'} );
}

my $s = new SC6::Cam::Sun();
print "Now: $dt\n" if ( $debug );
if ( $directory ) {
    print scalar localtime(), " We hav a direcotry to process: $directory\n";
}
elsif ( $force ) {
    print scalar localtime(), " We're forced to do this\n" if ( $debug );
}
else {
    if ( $s->is_sun($dt) ) {
        print scalar localtime(), " Sun is up!\n" unless ( $silent);
    }
    elsif ( $s->is_hour_after_dusk($dt) ) {
        print scalar localtime(), " Sun is down, but for less than an hour!\n" unless ( $silent);
    }
    else  {
        print scalar localtime(), " Sun is down!\n" unless ( $silent);
        exit;
    }
}

if ( $dryrun ) {
    print "This is a dry run, not doing anything\n" unless ( $silent);
}

my $format = 'orig';
make_moovie($format, $mode);
my %metadata = collect_metadata();
compress_moovie($format, $mode, %metadata);
my $push_return_code;
if ( ! $no_push ) {
    $push_return_code = push_to_youtube($format, $mode);
}
cleanup($format, $mode);

if ( $debug ) {
    print scalar localtime(), " buh bye\n";
    print scalar localtime(), " push still returned $push_return_code\n";
}
if ( $push_return_code != 0 ) {
    exit 1;
}
exit($push_return_code);

sub make_moovie {
    my ($format, $mode) = @_;

    my $image_dir = 'public';
    my $out = get_video_file($dt, $format, 'avi', $mode);

    if ( $directory ) {
        $image_dir = $directory;
    }
    else {
        $image_dir = get_image_dir($dt, $image_dir, $mode);
    }

    if ( ! $fps ) {
        $fps = $config->{'FPS'};
    }

    my $in = "'mf://" . $image_dir . "/*_" . $format . ".jpg'";
    my $mf = "w=" . $config->{'Sizes'}->{$format}->{'width'} . 
        ":h=" . $config->{'Sizes'}->{$format}->{'height'} . 
        ":type=" . $config->{'Type'} . 
        ":fps=" . $fps;

    my $cmd = "mencoder -msglevel all=1 -nosound -noskip -oac copy -ovc copy -o $out -mf $mf $in";
    my $ret = my_do_cmd($cmd, $dryrun);
    if ( $ret != 0 ) {
        print "Failed to make a movie, maybe no images?  Return: $ret\n";
        exit($ret);
    }
}

sub compress_moovie {
    my ($format, $mode, %metadata) = @_;
    my $in = get_video_file($dt, $format, 'avi', $mode);
    my $out = get_video_file($dt, $format, 'mp4', $mode);
    my $ll =  $config->{'FFMpegLogLevel'};
    my $md;
    foreach my $k ( keys %metadata ) {
	$md .= " -metadata $k='" . $metadata{$k} . "'";
    }
    my $cmd = $config->{'Bins'}->{'ffmpeg'};
    $cmd .= " -y -loglevel $ll -i $in $md -r 25 -s 1920x1080 -vcodec libx264 -b:v 30000k $out";
    my_do_cmd($cmd, $dryrun);
}

sub usage
{
    print "usage: $0 [-d|--debug] [-t|--date=date] [-f|--force] [-h|--help] [--silent] [--trickle] [-n|--dry-run] [-m|mode=mode]\n";
    print "\t-f|--force     - force building of the video files\n";
    print "\t-t|--date      - date of files to build and push\n";
    print "\t-D|--directory - directory to build and push (overrides date)\n";
    print "\t-h|--help      - this message\n";
    print "\t-n|--dry-run   - perform a trial run with no changes made\n";
    print "\t-k|--keep      - don't remove disposable video files\n";
    print "\t-np|--no-push  - don't push to youtube\n";
    print "\t--trickle      - use trickle to limit bandwidth\n";
    print "\t--fps          - override frame per second\n";
    print "\t--silent       - don't print normal amount of information\n";
    print "\t--debug        - print extra debugging information (debug overrides silent)\n";
    print "\t-m|--mode      - mode, prod or test\n";
    exit(1);

}

sub cleanup {
    my ($format, $mode) = @_;

    # don't delete video files if we were asked to keep them
    return if ( $keep );

    if ( $config->{'Video'}->{'Daily'}->{'Disposable'} ) {
        my $out = get_video_file($dt, $format, 'mp4', $mode);
        unlink $out or die "Can't remove $out: $!\n";
        $out = get_video_file($dt, $format, 'avi', $mode);
        unlink $out or die "Can't remove $out: $!\n";
    }
}

sub push_to_youtube {
    my ($format, $mode) = @_;

    my $cmd = "";
    if ( $trickle ) {
        $cmd = $trickle_cmd . " ";
    }
    $cmd .= $config->{'Bins'}->{'push2youtube'} . " " . $config->{'Bins'}->{'push2youtube_args'};
    if ( $date_from_args == 1 ) {
        $cmd .= " " . "--videoDate=" . $dt->ymd;
    }
    my $out = get_video_file($dt, $format, 'mp4', $mode);
    $cmd .= " --file " . $out;
    my $rcode = my_do_cmd($cmd, $dryrun);
    return $rcode;
}

sub collect_metadata {
    my %h;
    $h{'title'} = "test title";
    $h{'tags'} = "test tags";
    $h{'description'} = "test description";
    $h{'date'} = "test date";

    return %h;
}

sub my_do_cmd {
    my ($cmd, $dryrun) = @_;
    if ( $debug ) {
        print scalar localtime(), " ", $cmd, "\n";
    }
    if ( ! $dryrun ) {
        print scalar localtime(), " Start command\n" if ( $debug );
        my  $ret = `$cmd 2>&1`;
        print $ret if ( $debug );
        print scalar localtime(), " Stop command\n" if ( $debug );
        return $?;
    }
}

sub new_do_cmd {
    my $cmd = shift;
    my $wtr = gensym;
    my $rdr = gensym;
    my $err = gensym;

    print TMPOUT "command: $cmd\n";

    my $errors;
    if( ! open3($wtr, $rdr, $err, $cmd) ) {
        print TMPOUT "open3 of cmd $cmd failed $!";
        $errors = 1;
        return 1;
    }
    close $wtr;
    while(<$rdr>) {
        print TMPOUT;
    }
    close $rdr;
    while(<$err>) {
        print TMPOUT;
        $errors = 1;
    }
    close $err;

    return 0;
}


__DATA__
This exists so flock() code above works.
DO NOT REMOVE THIS DATA SECTION.

File: bin/build_time_lapse.py
#!/usr/bin/python3

#  todo
#  - Concurrance locking
#  - deal with building / pushing a date other than today
#  - make collect_metadata meaningful
#  - trickle, do we need trickle?

import sys
import os
import fcntl
from datetime import *
import argparse
import yaml
import errno
import logging
import logging.config
import pprint

sys.path.append('/usr/local/cam/lib/pythonlib')
import sc6_bucket_shiz
import sc6_sun
import sc6_general
import sc6_config


def today_video_name():
    vdir = config['BucketShiz']['VideoDir']
    t_name = config['BucketShiz']['TodayVideoName']
    out = vdir + "/" + t_name
    return(out)


def get_dest_name():
    vdir = config['BucketShiz']['VideoDir']
    dt = datetime.today()
    fname_template = config['BucketShiz']['VideoNameTemplate']
    fname_w_date = dt.strftime(fname_template)
    suffix = config['BucketShiz']['VideoNameSuffix']
    out = "%s/%s/%s/%s/%s%s" % (vdir,
                                dt.strftime("%Y"),
                                dt.strftime("%m"),
                                dt.strftime("%d"),
                                fname_w_date,
                                suffix)
    return(out)


def set_cache_timeout(dest_name):
    fbase = os.path.basename(dest_name)
    cache_control = config['BucketShiz']['Cache']['CacheControl']
    try:
        cache_timeout = config['BucketShiz']['Cache']['MaxAge'][fbase]
    except KeyError:
        cache_timeout = \
            config['BucketShiz']['Cache']['MaxAge']['default_video']
    cachecontrol_val = "'" + cache_control + ", max-age=" + \
        str(cache_timeout) + "'"
    bshiz.set_blob_cachecontrol(dest_name, cachecontrol_val)


class Vmeta:
    def __init__(self):
        self.title = "test title"
        self.tags = "test tags"
        self.description = "test description"
        self.date = "test date"


def collect_metadata():
    h = {}
    h['title'] = "test title"
    h['tags'] = "test tags"
    h['description'] = "test description"
    h['date'] = "test date"

    return h


def make_moovie(dt, size, mode, args):
    image_dir = 'public'
    out = sc6_general.get_video_file(dt, size, 'avi', mode)

    if args.directory:
        image_dir = args.directory
    else:
        image_dir = sc6_general.get_image_dir(dt, image_dir, mode)

    if args.fps:
        fps = args.fps
    else:
        fps = config['FPS']

    files_in = 'mf://{}/*_{}.jpg'.format(image_dir, size)
    mf = 'w={}:h={}:type={}:fps={}'.format(config['Sizes'][size]['width'],
                                           config['Sizes'][size]['height'],
                                           config['Type'],
                                           fps)

    cmd = ["mencoder", "-msglevel", "all=1",
           "-nosound", "-noskip", "-oac", "copy",
           "-ovc", "copy", "-o", out, "-mf", mf, files_in]
    logger.debug(" ".join(cmd))

    if not args.dryrun:
        return(sc6_general.run_cmd(cmd, debug))


def compress_moovie(dt, size, mode, metadata, args):
    infile = sc6_general.get_video_file(dt, size, 'avi', mode)
    outfile = sc6_general.get_video_file(dt, size, 'mp4', mode)

    cmd = [config['Bins']['ffmpeg'], "-y", "-loglevel",
           config['FFMpegLogLevel'], "-i", infile]
    for k in metadata.keys():
        cmd.append("-metadata")
        cmd.append(k + "='" + metadata[k] + "'")
    the_rest = ["-r", "25", "-s", "1920x1080", "-vcodec",
                "libx264", "-b:v", "30000k", outfile]
    cmd = cmd + the_rest

    if debug:
        print(" ".join(cmd))
    logger.debug(" ".join(cmd))

    if not args.dryrun:
        return(sc6_general.run_cmd(cmd, debug))


def cleanup(size, mode, args):

    # don't delete video files if we were asked to keep them
    if args.keep:
        return

    if config['Video']['Daily']['Disposable']:
        out = sc6_general.get_video_file(dt, size, 'mp4', mode)
        try:
            os.unlink(out)
        except OSError as error:
            print("Can't remove {}: {}".format(out, error))
            pass
        out = sc6_general.get_video_file(dt, size, 'avi', mode)
        try:
            os.unlink(out)
        except OSError as error:
            print("Can't remove {}: {}".format(out, error))
            pass


if __name__ == '__main__':

#  lock
#  try:
#         lokky = open(lock_file, 'w+')
#         fcntl.flock(lokky, fcntl.LOCK_EX | fcntl.LOCK_NB)
#         break
#  except IOError as e:
#         # raise on unrelated IOErrors
#         if e.errno != errno.EAGAIN:
#             raise
#         else:
#
#             time.sleep(0.1)
#  fcntl.flock(x, fcntl.LOCK_EX | fcntl.LOCK_NB)

    argparser = argparse.ArgumentParser(description='build time lapse video.')
    argparser.add_argument("--file", dest="file",
                           required=False,
                           help="Video file to upload")
    argparser.add_argument("--directory", dest="directory",
                           default=None,
                           help="a directory")
    argparser.add_argument("--fps", dest="fps", help="frames per second")
    argparser.add_argument("--mode", dest="mode", help="prod, test, dev",
                           default="prod")
    argparser.add_argument("--date", dest="date_in", help="override date")
#  how we implement something like trickle and throttleing?
#     argparser.add_argument("--trickle", dest="trickle",  action='store_true',
#                            default=False,
#                            help="trickle the data in")
#     argparser.add_argument("--trickle-cmd", dest="trickle_cmd",
#                            default="trickle -s -u 200",
#                            help="trickle the data in")
    argparser.add_argument("--dryrun", action='store_true', dest='dryrun',
                           help="don't actually do anything")
    argparser.add_argument("--force", action='store_true', dest='force',
                           help="force thigns to happen")
    argparser.add_argument("--videoDate", dest="videoDate", default="",
                           help="video is for some other date than today")
    argparser.add_argument("--keep", action='store_true',
                           default=False,
                           help="keep video, don't delete")
    argparser.add_argument("--silent", action='store_true', dest='silent',
                           default=False,
                           help="no output")
    argparser.add_argument("--debug", action='store_true', dest='debug',
                           help="lots of output",
                           default=False)
    argparser.add_argument("--dontUpload", action='store_true',
                           help="for testing, don't actually do the upload")
    args = argparser.parse_args()

    cfg = sc6_config.Config(mode = args.mode)
    config = cfg.get_config()

    #  create logger
    logger = logging.getLogger('build_timelapse')

    # either from the cmd line or config    
    debug = args.debug or config['Debug']

    # debug trumps silent
    if debug and args.silent:
        args.silent = False

    if debug:
        print("Debug output on")

    mysun = sc6_sun.SC6Sun(config)
    dt = mysun.get_dt()

    # titles and stuff for video
    if args.videoDate:
        date_string = args.videoDate
    else:
        date_string = dt.strftime("%Y%m%d")

#    date_from_args = False
#    if date:
#        if debug:
#            print date, "\n"
#        my (seconds, error) = parsedate($date)
#        if ! seconds:
# 	   print "Error can't parse date: $date : $error\n"
#            exit(1)
#        else:
# 	   print seconds, "\n" if ( debug )
#            dt = DateTime->from_epoch(
#                    epoch => $seconds, time_zone =>
#                    config['General']['Timezone'] )
#        force = 1
#        date_from_args = 1
#    else:
#        dt = mysun.get_dt()

    if debug:
        print("Now: ", dt)

    if args.directory:
        print(" We hav a direcotry to process: ", args.directory)
        # maybe check the directory
        # specified directory implies force
        args.force = True
    elif args.force:
        if debug:
            print(" We're forced to do this")
    elif mysun.is_sun():
        if debug:
            print(" Sun is up!")
    elif mysun.is_hour_after_dusk():
        if debug:
            print(" Sun is down, but for less than an hour!")
    else:
        if debug:
            print(" Sun is down!")
        sys.exit()

    if args.dryrun and not args.silent:
        print("This is a dry run, not doing anything")

    # make sure we have the directory we're going to work with
    if args.directory:
        image_dir = args.directory
    else:
        image_dir = sc6_general.get_image_dir(dt, "public", args.mode)
    if not os.path.exists(image_dir) or len(os.listdir(image_dir) ) == 0:
        print("directory {} missing or empty, quitting".format(image_dir))
        sys.exit()

    size = 'orig'
    make_moovie(dt, size, args.mode, args)

    metadata = collect_metadata()
    compress_moovie(dt, size, args.mode, metadata, args)

    bshiz = sc6_bucket_shiz.MyBucket(config)

    logger.debug("upload logic")
    if args.file:
        if not os.path.exists(args.file):
            exit("Please specify a valid file using the --file= parameter.")
        vid_file = args.file
    else:
        vid_file = sc6_general.get_video_file(dt, size, 'mp4', args.mode)
        dest_name = get_dest_name()
        if debug:
            print(dest_name)

        if args.dontUpload or args.dryrun:
            logger.info("dontUpload is set, we won't upload")
        else:
            # send-up the video and set cache
            logger.info("upload {} to {}".format(vid_file, dest_name))
            bshiz.upload_blob(vid_file, dest_name)
            set_cache_timeout(dest_name)

            today_name = today_video_name()
            bshiz.cp_in_bucket(dest_name, today_name)
            logger.info("going to copy {} to {}".format(dest_name, today_name))
            set_cache_timeout(today_name)

    cleanup(size, args.mode, args)

    #     # release lock
    #     fcntl.flock(x, fcntl.LOCK_UN)
